scripts/apps/authoring/attachments/reducer.ts

```diff
@@ -0,0 +1,32 @@
+import {
+    INIT_ATTACHMENTS,
+    ADD_ATTACHMENTS,
+    REMOVE_ATTACHMENT,
+    EDIT_ATTACHMENT,
+    SAVE_ATTACHMENT,
+} from './actions';
+
+const initialState = {
+    diff: null,
+    file: null,
+    files: [],
+    maxSize: 0,
+    maxFiles: 10,
+};
+
+export function attachments(state = initialState, action) {
+    switch (action.type) {
+    case INIT_ATTACHMENTS:
+        return {...action.payload};
+    case ADD_ATTACHMENTS:
+        return {...state, files: state.files.concat(action.payload)};
+    case REMOVE_ATTACHMENT:
+        return {...state, files: state.files.filter((f) => f !== action.payload)};
+    case EDIT_ATTACHMENT:
+        return {...state, file: action.payload, diff: action.payload != null ? Object.create(action.payload) : null};
```

Comment:
Why are you using `Object.create` in ` Object.create(action.payload)`?

Reply:
that's like a diff, so that object will only have own properties for modified keys

Reply:
I'm not sure I get that. When you do `Object.create` you simply create a copy of an object. Is there more to it? How is it related to diffs?

Reply:
yep so you can use new object to access old properties, it will get it via inheritance, but once you change it there will be an own property created on the diff one
```
const original = {foo: 1};
const diff = Object.create(original);

diff.foo
> 1

diff.hasOwnProperty('foo')
> false

diff.foo = 2;

diff.hasOwnProperty('foo')
> true
```
so when we use it later for patch request it's only sending own properties of diff, so only parts that were changed.

Reply:
That's a neat trick, I haven't seen diffing done this way before. As you can see from me not understanding this right away, it would be good instead of calling `object.create` to call `prepareDiff` which would call `object.create` except that it would be documented in the function body how the diffing works.

Why are we calculating the diffs in the first place? Don't etags protect from conficting writes? Do etags work on document level or it takes the fields into account too?

Reply:
that's not about etags and conflicting writes, but to reduce payloads and not to send data which were not changed.

Reply:
I'm all for saving the data, but I'd like it to be abstracted away properly. Can it be done in a way that reducer and the whole application except the function which doing the diffing is not required to know about it? If so, we should do it.

Reply:
at the end I got rid of it, handling those changes as component state